lib/postgrex/extensions/tsvector.ex

```diff
@@ -0,0 +1,127 @@
+defmodule Postgrex.Extensions.Tsvector do
+  @moduledoc false
+
+  import Postgrex.BinaryUtils, warn: false
+  use Bitwise
+  use Postgrex.BinaryExtension, [send: "tsvectorsend"]
+
+  def encode(_) do
+    quote location: :keep do
+      values when is_list(values) ->
+        encoded_tsvector = unquote(__MODULE__).encode_tsvector(values)
+        <<byte_size(encoded_tsvector) :: int32, encoded_tsvector :: binary>>
+      other ->
+        raise ArgumentError,
+        Postgrex.Utils.encode_msg(other, "a list of tsvectors")
+    end
+  end
+
+  def decode(_) do
+    quote do
+      <<len :: int32, value :: binary-size(len)>> ->
+        <<nb_lexemes :: int32, words :: binary>> = value
+        unquote(__MODULE__).decode_tsvector_values(:binary.bin_to_list(words))
+    end
+  end
+
+  ## Helpers
+
+  def encode_tsvector(values) do
+    << length(values) :: int32, encode_lexemes(values) :: binary >>
+  end
+
+  defp encode_lexemes(values) do
+    Enum.map(values, fn(x) ->
+        case x do
+          %Postgrex.Lexeme{word: word, positions: positions} ->
+            %{word: "#{word}\0", positions: positions}
+          end
+        end)
+      |> Enum.map(fn(x) -> encode_positions(x[:word], x[:positions]) end)
+      |> Enum.reduce("", fn(x, acc) -> acc <> x end)
+  end
+
+  defp encode_positions(bytes, position) do
+      positions = Enum.map(position, fn(x) -> << encode_weight(x)::2, encode_position(x)::14 >> end)
+                  |> Enum.reduce("", fn(x, acc) -> acc <> x end)
```

Comment:
This is a good improvement.

I was able to remove all the Enum.reduce in this file as they were all doing the same operation 👍 .